TextChoice Test Components and Automation

labkey-danield opened the pull request:
#### Rationale
Recently there was a new field type, TextChoice, added to LKS. The field is similar to a LookUp field, but the difference is that the allowed values are defined in the field and not in some other table. The way this is done is that the allowed values are implemented as a validator on the field (the field is basically a text type). The fact that a validator is used is hidden from the user. For a TextChoice field a user can only define the list of values, they do not get to give the validator a name, description or error message. No other validator can be applied to a TextChoice field.

A TextChoice field looks like this:
<img width="1079" alt="Screen Shot 2022-04-14 at 11 08 59 AM" src="https://user-images.githubusercontent.com/12738200/163454697-1b477e73-0e19-4866-9b44-3db75c2e52c9.png">

A user provides the list of allowed values by clicking the 'Add Values' button and then entering the values into the dialog that is shown:
<img width="602" alt="Screen Shot 2022-04-14 at 11 08 05 AM" src="https://user-images.githubusercontent.com/12738200/163454908-32721b44-1786-4a07-b6a5-74d4eb784d7f.png">

If a value has been used by an entry the value is locked (notice locked icon on top image) and cannot be deleted. However a locked field can be updated:
<img width="1065" alt="Screen Shot 2022-04-14 at 11 37 14 AM" src="https://user-images.githubusercontent.com/12738200/163455235-234936e9-8203-4d20-9682-5bdecf5daeb5.png">

If a field value is updated/changed to a value that already exists an error is shown (see top image).

If creating a sample type with a TextChoice field through the APIs you will need to create a FieldDefinition object and then pass the list of values into the setTextChoiceValues method:
```
        SampleTypeDefinition sampleTypeDefinition = new SampleTypeDefinition(sampleType);

        List<FieldDefinition> fields = new ArrayList<>();

        FieldDefinition textChoiceField = new FieldDefinition("Text_Choice", FieldDefinition.ColumnType.TextChoice);
        textChoiceField.setTextChoiceValues(List.of("A", "B", "C", "\u00DC"));
        fields.add(textChoiceField);

        sampleTypeDefinition.setFields(fields);

        SampleTypeAPIHelper.createEmptySampleType(getCurrentContainerPath(), sampleTypeDefinition);

```
TextChoice fields can appear in sample types and assay designs.

#### Related Pull Requests
* https://github.com/LabKey/sampleManagement/pull/939
* https://github.com/LabKey/biologics/pull/1276

#### Changes
* Add TextChoice UI components parts.
* Add a TextChoiceValidator to FieldDefinitions.
* Add test cases for sample types. Validates the TextChoice field in the designer.
* Add test cases for assays. Validates locking and updating.
* Added test cases for issues and lists. Just sanity test to validate TextChoice in other domain types.
* Added test case for exporting and then importing a folder with domains that have TextChoice fields.

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Update FieldDefinition to include TextChoice field type.
Update Domain UI components (DomainFormPanel & DomainFieldRow) to interact with TextChoice field types.
Add a dialog for adding TextChoice selections.
Fix CreateSamplesPage to be smarter about waiting.

labkey-danield committed:
Add an update method for locked value that checks for info message.

labkey-tchad commented on src/org/labkey/test/components/domain/DomainFieldRow.java:
```diff
@@ -600,6 +602,164 @@ public DomainFieldRow clickRemoveOntologyConcept()
+    /**
+     * Update a TextChoice value that is currently being used and expect an info message with the number updated.
+     *
+     * @param originalValue The original value to select and change.
+     * @param newValue The new value to replace it.
+     * @return The info message indicating the number of entites updated.
+     */
+    public String updateLockedTextChoiceValue(String originalValue, String newValue)
+    {
+        updateTextChoiceValue(originalValue, newValue);
+        WebElement alert = BootstrapLocators.infoBanner.waitForElement(this, 1_000);
+        return alert.getText();
+    }
+
+    /**
+     * Check to see if the 'Delete' button is enabled for a TextChoice field. Note: The button is only visible if a
+     * value is selected.
+     *
+     * @return True if the button is enabled, false otherwise.
+     */
+    public boolean isTextChoiceDeleteButtonEnabled()
+    {
+        WebElement button = Locator.button("Delete").findElement(this);
+        return button.isEnabled();
+    }
+
+    /**
+     * Delete a value from the list of values for a TextChoice field. This will select the value before deleting it.
+     *
+     * @param value The value to delete.
+     * @return This field row.
+     */
+    public DomainFieldRow deleteTextChoiceValue(String value)
+    {
+        selectTextChoiceValue(value);
+        WebElement button = Locator.button("Delete").findElement(this);
+        WebDriverWrapper.waitFor(button::isEnabled, "'Delete' button is not enabled.", 1_000);
+        button.click();
+        return this;
+    }
```
Should this wait for the option to go away? Maybe wait for the Delete button to be disabled after clicking.

labkey-danield replied:
Should be instantaneous, but to be safe I added a wait for the delete button to disappear (won't be disabled because no value will be selected after the selected value is deleted).

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-tchad commented on src/org/labkey/test/components/domain/DomainFormPanel.java:
```diff
@@ -120,6 +121,15 @@ else if (validator instanceof FieldDefinition.RangeValidator)
                 {
                     rangeValidators.add((FieldDefinition.RangeValidator) validator);
                 }
+                else if (validator instanceof FieldDefinition.TextChoiceValidator)
+                {
+                    // TextChoice is a field type and not an attribute on a field, any other validators that are applied
+                    // will be ignored (TextChoice field cannot have validators).
+                    textChoiceValidator = (FieldDefinition.TextChoiceValidator) validator;
+                    regexValidators = new ArrayList<>();
+                    rangeValidators = new ArrayList<>();
+                    break;
+                }
```
Should just error if there are multiple validators instead of ignoring them silently.
```suggestion
                else if (validator instanceof FieldDefinition.TextChoiceValidator textChoiceValidator)
                {
                    // TextChoice is a field type; implemented using a special validator. TextChoice field cannot have other validators.
                    if (validators.size() > 1)
                    {
                        throw new IllegalArgumentException("TextChoice fields cannot have additional validators.");
                    }
                    fieldRow.setTextChoiceValues(textChoiceValidator.getValues());
                }
```

labkey-danield replied:
Fair enough. Suggested change applied.

labkey-tchad commented on src/org/labkey/test/params/FieldDefinition.java:
```diff
@@ -384,6 +385,13 @@ public FieldDefinition setPrincipalConceptSearchExpression(String ontologyName,
         return this;
     }
 
+    public FieldDefinition setTextChoiceValues(List<String> values)
+    {
+        Assert.isTrue(getType().equals(ColumnType.TextChoice), "Type is not set to 'ColumnType.TextChoice'.");
+        setValidators(List.of(new FieldDefinition.TextChoiceValidator(values)));
+        return this;
+    }
```
Non-standard assert.
```suggestion
    public FieldDefinition setTextChoiceValues(List<String> values)
    {
        Assert.assertEquals("Invalid field type for text choice values.", ColumnType.TextChoice, getType());
        setValidators(List.of(new FieldDefinition.TextChoiceValidator(values)));
        return this;
    }
```

labkey-danield replied:
Oops! Updated.

labkey-tchad commented on src/org/labkey/test/params/FieldDefinition.java:
```diff
@@ -836,4 +846,47 @@ public String getSecondRange()
+     */
+    public static class TextChoiceValidator extends FieldValidator<TextChoiceValidator>
+    {
+        private List<String> _values;
+
+        public TextChoiceValidator(List<String> values)
+        {
+            // The TextChoice validator only has a name and no description or message.
+            // And the name is generated (not user defined).
+            super("Text Choice Validator", "", "");
+            _values = values;
+        }
```
Consider locking down the collection of values.
```suggestion
        private final List<String> _values;

        public TextChoiceValidator(List<String> values)
        {
            // The TextChoice validator only has a name and no description or message.
            // And the name is generated (not user defined).
            super("Text Choice Validator", "", "");
            _values = Collections.unmodifiableList(values);
        }
```

labkey-danield replied:
Done.

labkey-danield committed:
Add search for TextChoice values to DomainFieldRow.
Add test cases for LKS sample type.

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Add test cases for LKS sample type.
Implemented code review feedback for components.
Added SampleTypeHelper.mergeImportExpectingError method.

labkey-danield committed:
Fixing some bugs.

labkey-danield committed:
Adding test for Assays.

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Adding test for Assays.

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Adding test for Import/Export.
Move some shared code into a base abstract class.

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Added test to validate TextChoice in an issue design.
Moved other domain tests into the import/export tests.
Added comments.

labkey-danield committed:
Added a TextChoice test for LKB.
Adding check for field name casing based on DB type.

labkey-danield committed:
Adding some comments.

cnathe commented on src/org/labkey/test/tests/TextChoiceImportExportAndOtherDomainsTest.java:
```diff
@@ -0,0 +1,376 @@
+    /**
+     * Create a run for the assay.
+     */
+    protected void createAssayRun()
```
how similar is this to the `createAndValidateAssayRun` method in the TextChoiceAssayTest? any reason for looking to share the assay run import parts?

labkey-danield replied:
Yep, had thought about that. Moved the create run code out to a shared method.

labkey-tchad commented on src/org/labkey/test/tests/TextChoiceSampleTypeTest.java:
```diff
@@ -0,0 +1,653 @@
+
+        List<String> lockedValues = fieldRow.getLockedTextChoiceValues();
+
+        // Sort the result from the UI, at this point more concerned about the values and not the order.
+        Collections.sort(lockedValues);
+        Collections.sort(valuesUsed);
+
+        checker()
+                .withScreenshot("ST_Designer_Locked_Values_Error")
+                .verifyEquals("Locked values not as expected.", valuesUsed, lockedValues);
```
Could use the new `assertEqualsSorted` method here (and below):
```suggestion

        checker()
                .withScreenshot("ST_Designer_Locked_Values_Error")
                .wrapAssertion(() -> LabKeyAssert.assertEqualsSorted("Locked values not as expected.", valuesUsed, lockedValues));
```

Might even be worth creating `DeferredErrorCollector.verifyEqualsSorted` instead of using `wrapAssertion`.

labkey-tchad commented on src/org/labkey/test/tests/TextChoiceSampleTypeTest.java:
```diff
@@ -0,0 +1,653 @@
+
+        // Some TextChoice values.
+        List<String> expectedUnLockedValues = new ArrayList<>();
+        expectedUnLockedValues.add("ÅÅ");
```
Use `\u` syntax for special characters in java Strings.

labkey-danield replied:
Yep, missed it.

labkey-tchad commented on src/org/labkey/test/tests/TextChoiceSampleTypeTest.java:
```diff
@@ -0,0 +1,653 @@
+        expectedUnLockedValues.add("CC");
+        expectedUnLockedValues.add("DD");
+        expectedUnLockedValues.add("E E E");
+        expectedUnLockedValues.add("ƒƒ");
```
`\u` encode these.

labkey-danield replied:
And missed it here as well.

labkey-tchad reviewed (approved):
Only big problem I see is the unicode characters not using `\u` encoding in some strings.

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Update src/org/labkey/test/tests/TextChoiceSampleTypeTest.java

Co-authored-by: Trey Chadick <tchad@labkey.com>

labkey-danield committed:
Escaping extended ascii characters used in test data.
Add test for validating TextChoice in Sources.
Moved shared assay code into the TextChoiceTest base class.

labkey-danield committed:
Merge branch 'develop' into fb_textChoiceAutomation

labkey-danield committed:
Updating no-op checks in QueryGrid.
